examples/restriction/non_topologically_sorted_functions/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+
+        let mut must_come_before: HashSet<(LocalDefId, LocalDefId)> = HashSet::new();
+        // stores all call sites for (caller, callee)
+        let mut call_sites: HashMap<(LocalDefId, LocalDefId), Vec<Span>> = HashMap::new();
+
+        for caller_id in def_order {
+            // use hir_maybe_body_owned_by — works for functions and methods etc.
+            let caller_body = cx.tcx.hir_maybe_body_owned_by(caller_id);
+
+            if let Some(caller_body) = caller_body {
+                let caller_body_id = caller_body.id();
+                let callees: Vec<Callee> = Self::collect_callees_in_body(cx, caller_body_id);
+
+                must_come_before = Self::build_caller_callee_constraint(
+                    caller_id,
+                    &callees,
+                    must_come_before,
+                    &mut call_sites,
+                );
+                must_come_before = Self::build_multiple_precedence_rule(&callees, must_come_before);
+            }
+        }
+
+        let violations = Self::find_violations(cx, &must_come_before, &functions);
+        let mut warned: HashSet<LocalDefId> = HashSet::new();
+
+        for violation in violations {
+            let Violation {
+                name_first_fn,
+                name_second_fn,
+                id_first_fn,
+                id_second_fn,
+                span,
+                ..
+            } = violation;
+            if warned.insert(id_first_fn) {
+                cx.span_lint(NON_TOPOLOGICALLY_SORTED_FUNCTIONS, span, |diag| {
+                    // PRIMARY ERROR
+                    diag.span_label(
+                        span,
+                        format!(
+                            "function `{name_first_fn}` should be defined before `{name_second_fn}`"
+                        ),
+                    );
+
+                    diag.help(
+        "move the function earlier in the module so callers and callee ordering is respected",
+    );
+
+                    // search call sites
+                    if let Some(sites) = call_sites.get(&(id_first_fn, id_second_fn)) {
+                        let sm = cx.sess().source_map();
+                        let good_sites: Vec<Span> = sites
+                            .iter()
+                            .copied()
+                            .filter(|s| !s.from_expansion() && !s.in_external_macro(sm))
+                            .collect();
+
+                        if let Some(call_span) = good_sites.first() {
+                            // НОВОЕ: extra block with double info
+                            diag.span_note(
+                                *call_span,
+                                format!("`{name_second_fn}` is called from `{name_first_fn}` here"),
+                            );
+                        }
+                    } else {
+                        // if the second rule
+                        diag.note("no call site recorded (internal)");
+                    }
```

Comment:
```suggestion
                    }
```
Let's please remove this `else` block. I don't think it is necessary.